File: php/bob/bob.php
<?php

class Bob
{
    public function respondTo($ask)
    {
        $ask = trim($ask);
        if (substr($ask, -1) === '?') {
            if ($this->checkUppercase($ask)) {
                return 'Calm down, I know what I\'m doing!';
            }
            return 'Sure.';
        } else if($this->checkUppercase($ask)) {
            return 'Whoa, chill out!';
        } else if ($ask === '') {
            return 'Fine. Be that way!';
        } else {
            return 'Whatever.';
        }
    }

    protected function checkUppercase($string)
    {
        $string = preg_replace('/[^a-z]+/i', '', $string);
        return ctype_upper($string);
    }
}

File: dart/bob/lib/bob.dart
class Bob {
  bool _checkUppercase(String question) {
    return question == question.toUpperCase() &&
        question.contains(new RegExp(r'[A-Z]'));
  }

  String response(String question) {
    question = question.replaceAll(RegExp(r"\s+"), '');
    if (question.isEmpty) {
      return 'Fine. Be that way!';
    }
    if (question.substring(question.length - 1) == '?') {
      if (_checkUppercase(question))
        return 'Calm down, I know what I\'m doing!';
      else
        return 'Sure.';
    }
    if (_checkUppercase(question)) return 'Whoa, chill out!';

    return 'Whatever.';
  }
}
